README.txt
/* $Id: README.txt,v 1.1.2.1.2.25.2.9 2010/07/25 16:33:19 pwolanin Exp $ */

This module integrates Drupal with the Apache Solr search platform. Solr search
can be used as a replacement for core content search and boasts both extra
features and better performance. Among the extra features is the ability to have
faceted search on facets ranging from content author to taxonomy to arbitrary
CCK fields.

The module comes with a schema.xml and solrconfig.xml file which should be used
in your Solr installation.

This module depends on the search framework in core. However, you may not want
the core searches and only want Solr search. If that is the case, you want to
use the Core Searches module in tandem with this module.

When used in combination with core search module, Apache Solr is not the default
search. Access it via a new tab on the default search page, called "Search".

Installation
------------

Prerequisite: Java 5 or higher (a.k.a. 1.5.x).  PHP 5.1.4 or higher.

Those with PHP < 5.2.0 must install the PECL json module or download
the Json code from the Zend Framework (see below).

Install the Apache Solr Drupal module as you would any Drupal module.

Before enabling it, you must also do the following:

Get the PHP library from the external project. The project is
found at:  http://code.google.com/p/solr-php-client/
From the apachesolr module directory, run this command:

svn checkout -r22 http://solr-php-client.googlecode.com/svn/trunk/ SolrPhpClient

Alternately you may download and extract the library from
http://code.google.com/p/solr-php-client/downloads/list

Make sure to select a r22 archive, either of these two:
http://solr-php-client.googlecode.com/files/SolrPhpClient.r22.2009-11-09.zip
http://solr-php-client.googlecode.com/files/SolrPhpClient.r22.2009-11-09.tgz

Note that revision 22 is the currently tested and required revision. 
Make sure that the final directory is named SolrPhpClient under the apachesolr
module directory.

If you are maintaing your code base in subversion, you may choose instead to 
use svn export or svn externals. For an export (writing a copy to your local
directory without .svn files to track changes) use:

svn export -r22 http://solr-php-client.googlecode.com/svn/trunk/ SolrPhpClient

Instead of checking out, externals can be used too. Externals can be seen as 
(remote) symlinks in svn. This requires your own project in your own svn ]
repository, off course. In the apachesolr module directory, issue the command:

svn propedit svn:externals .

Your editor will open. Add a line

SolrPhpClient -r22 http://solr-php-client.googlecode.com/svn/trunk/

On exports and checkouts, svn will grab the externals, but it will keep the 
references on the remote server.

Those without svn, etc may also choose to try the bundled Acquia Search
download, which includes all the items which are not in Drupal.org CVS due to 
CVS use policy. See the download link here: 
http://acquia.com/documentation/acquia-search/activation

Download Solr 1.4 from:
http://www.apache.org/dyn/closer.cgi/lucene/solr/

Unpack the tarball somewhere not visible to the web (not in your apache docroot
and not inside of your drupal directory).

The Solr download comes with an example application that you can use for
testing, development, and even for smaller production sites. This
application is found at apache-solr-nightly/example.

Move apache-solr-nightly/example/solr/conf/schema.xml and rename it to
something like schema.bak. Then move the schema.xml that comes with the
ApacheSolr Drupal module to take its place.

Similarly, move apache-solr-nightly/example/solr/conf/solrconfig.xml and rename
it like solrconfig.bak. Then move the solrconfig.xml that comes with the
ApacheSolr Drupal module to take its place.

Now start the solr application by opening a shell, changing directory to
apache-solr-nightly/example, and executing the command java -jar start.jar

Test that your solr server is now available by visiting
http://localhost:8983/solr/admin/

For those using PHP 5.1, you must either install the PECL json extension
into PHP on your sever, or you may use the Zend framework Json library.
for the PECL extension see:  http://pecl.php.net/package/json
The Solr client has been tested with Zend framework release 1.7.7.
To get this code, you may use svn from the apachesolr directory:
svn co http://framework.zend.com/svn/framework/standard/tags/release-1.7.7/library/Zend
However, the only required parts are:
http://framework.zend.com/svn/framework/standard/tags/release-1.7.7/library/Zend/Exception.php
http://framework.zend.com/svn/framework/standard/tags/release-1.7.7/library/Zend/Json/
The 'Zend' directory should normally be under the apachesolr
directory, but may be elsewhere if you set that location to be
in your PHP include path.

Now, you should  enable the "Apache Solr framework" and "Apache Solr search" 
modules. Check that you can connect to Solr at ?q=admin/setting/apachesolr
Now run cron on your Drupal site until your content is indexed. You
can monitor the index at ?q=admin/settings/apachesolr/index

The solrconfig.xml that comes with this modules defines auto-commit, so
it may take a few minutes between running cron and when the new content
is visible in search.

Enable blocks for facets first at Administer > Site configuration > Apache Solr > Enabled filters,
then position them as you like at Administer > Site building > Blocks.   

Configuration variables
--------------

The module provides some (hidden) variables that can be used to tweak its
behavior:

 - apachesolr_luke_limit: the limit (in terms of number of documents in the
   index) above which the module will not retrieve the number of terms per field
   when performing LUKE queries (for performance reasons).

 - apachesolr_tags_to_index: the list of HTML tags that the module will index
   (see apachesolr_add_tags_to_document()).

 - apachesolr_exclude_comments_types: an array of node types. Any type listed
   will have any attached comments excluded from the index.

 - apachesolr_ping_timeout: the timeout (in seconds) after which the module will
   consider the Apache Solr server unavailable.

 - apachesolr_optimize_interval: the interval (in seconds) between automatic
   optimizations of the Apache Solr index. Set to 0 to disable.

 - apachesolr_cache_delay: the interval (in seconds) after an update after which
   the module will requery the Apache Solr for the index structure. Set it to
   your autocommit delay plus a few seconds.

 - apachesolr_service_class: the Apache_Solr_Service class used for communicating
   with the Apache Solr server.

 - apachesolr_query_class: the default query class to use.

 - apachesolr_cron_mass_limit: update or delete at most this many documents in
   each Solr request, such as when making {apachesolr_search_node} consistent
   with {node}.

Troubleshooting
--------------
Problem:
Links to nodes appear in the search results with a different host name or
subdomain than is preferred.  e.g. sometimes at http://example.com
and sometimes at http://www.example.com

Solution:
Set $base_url in settings.php to insure that an identical absolute url is
generated at all times when nodes are indexed.  Alternately, set up a re-direct
in .htaccess to prevent site visitors from accessing the site via more than one
site address.


Developers
--------------

Exposed Hooks in 5.x:

hook_apachesolr_modify_query(&$query, &$params, $caller);

  Any module performing a search should call apachesolr_modify_query($query, $params, 'modulename'). 
  That function then invokes this hook. It allows modules to modify the query object and params array. 
  $caller indicates which module is invoking the hook.

  Example:

        function my_module_apachesolr_modify_query(&$query, &$params, $caller) {
          // I only want to see articles by the admin!
          $query->add_filter("uid", 1);         
        }        

CALLER_finalize_query(&$query, &$params);

  The module calling apachesolr_do_query() may implement a function that is run after
  hook_apachesolr_modify_query() and allows the caller to make final changes to the
  query and params before the query is sent to Solr.  The function name is built
  from the $caller parameter to apachesolr_do_query().

hook_apachesolr_prepare_query(&$query, &$params, $caller);

  This is pretty much the same as hook_apachesolr_modify_query() but runs earlier
  and before the query is statically cached. It can e.g. be used to add
  available sorts to the query.

  Example:

        function my_module_apachesolr_prepare_query(&$query) {
          // Add a sort on the node ID.
          $query->set_available_sort('nid', array(
            'title' => t('Node ID'),
            'default' => 'asc',
          ));
        }

hook_apachesolr_cck_fields_alter(&$mappings)

  Add or alter index mappings for CCK types.  The default mappings array handles just 
  text fields with option widgets:

    $mappings['text'] = array(
      'optionwidgets_select' => array('callback' => '', 'index_type' => 'string'),
      'optionwidgets_buttons' => array('callback' => '', 'index_type' => 'string')
    );

  In your _alter hook implementation you can add additional field types such as:

    $mappings['number_integer']['number'] = array('callback' => '', 'index_type' => 'integer');

  You can allso add a mapping for a specific field.  This will take precedence over any
  mapping for a general field type. A field-specific mapping would look like:

    $mappings['per-field']['field_model_name'] = array('callback' => '', 'index_type' => 'string');

  or

    $mappings['per-field']['field_model_price'] = array('callback' => '', 'index_type' => 'float');

hook_apachesolr_types_exclude($namespace)

  
  Invoked by apachesolr.module when generating a list of nodes to index for a given
  namespace.  Return an array of node types to be excldued from indexing for that namespace 
  (e.g. 'apachesolr_search'). This is used by apachesolr_search module to exclude 
  certain node types from the index.

hook_apachesolr_node_exclude($node, $namespace)

  This is invoked by apachesolr.module for each node to be added to the index - if any module
  returns TRUE, the node is skipped for indexing. 

hook_apachesolr_update_index(&$document, $node)

  Allows a module to change the contents of the $document object before it is sent to the Solr Server.
  To add a new field to the document, you should generally use one of the pre-defined dynamic fields. 
  Follow the naming conventions for the type of data being added based on the schema.xml file.

hook_apachesolr_search_result_alter(&$doc)

  The is invoked by apachesolr_search.module for each document returned in a search - new in 6.x-beta7
  as a replacement for the call to hook_nodeapi().

hook_apachesolr_sort_links_alter(&$sort_links)

  Called by the sort link block code. Allows other modules to modify, add or remove sorts.

Themers
----------------

See inline docs in apachesolr_theme and apachesolr_search_theme functions 
within apachesolr.module and apachesolr_search.module.



CHANGELOG.txt
// $Id: CHANGELOG.txt,v 1.1.2.99.2.135 2010/09/29 14:26:07 robertDouglass Exp $

Apache Solr integration x.x-x.x, xxxx-xx-xx
------------------------------

Apache Solr integration 6.x-2.x, xxxx-xx-xx
------------------------------
#925700 by janusman: Changed Quick fix: reduce t() string duplicity.
#910100 by robertDouglass: Changed Make whole $snippets array available to theme layer.

Apache Solr integration 6.x-2.0-BETA3, 2010-09-04
------------------------------
#840358 by pwolanin | torstenzenk: Fixed Error searching Taxonomies.
#828412 by eosrei | jpmckinney: Fixed apachesolr_get_facet_definition() searches incorrect field for definition.
#864146 by pwolanin | jpmckinney: Fixed When moving fq to q.alt, we should parenthesize each fq.
NEW MODULE: ApacheSolr Text; takes care of text fields from CCK with content permissions.
#704190 by pwolanin: Added Add page callbacks to display conf files in the index in reports.
#878996 by pwolanin | weri: Added Don't break the loop .
#373824 by pwolanin | David Stosik: Fixed Attachment count is not displayed in search results.
#896324 by ahankinson | jpmckinney: Fixed module_invoke() called too late.
#901720 by robertDouglass: Fixed Highlighting snippets in search results not flexible enough, and logic improvement.
#899590 by robertDouglass: Changed CCK facet definitions needs to expand its role and have a 'facets' param.
#899560 by robertDouglass: Fixed apachesolr_do_query() has to take $params by reference.
#751004 by pwolanin, jpmckinney, rjbrown99, robertDouglass | EugenMayer, camidoo: Fixed Respect NODE_BUILD_SEARCH_INDEX exclusion setting.
#885950 by pwolanin, csevb10 preserve added/removed filters when filterstring is re-parsed.
#880488 by pwolanin: security hardening: don't use unserialize on untrusted data.
#867448 by B-Prod | jpmckinney: Fixed Apache taxonomy search does not include the search.pages.inc file.
#864160 by pwolanin: Allow the caller a last chance to modify the query and params.
#847040 by Agileware: Allow other modules to use facet only searching.
#536990 by pwolanin | jpmckinney, janusman: always index content as an anonymous user.
#592522 by pwolanin, quaoar: use multi-table sytax for MySQL to avoid killer sub-selects.
#835850 by pwolanin: add more replicated files to solrconfig.xml master section.
#830976 by eosrei, pwolanin: make sure we return a non-zero ping time on success.
#824154 by pwolanin: update schema.xml to add preserveOriginal="1" to WordDelimiterFilterFactory.
#804590 by ygerasimov: Bug in building query from subqueries in Solr_Base_Query
#818608 by nofun: fix 5.3 warning on get_object_vars() 
#382878 by smoothify: Store the real teaser in the index


Apache Solr integration 6.x-2.0-BETA2, 2010-04-08
------------------------------
#660754 by jhedstrom: Added Allow key sorting of facets.
#614644 by netsensei | robertDouglass: Fixed Forms attached to Apachesolr search results won't work.
#747346 by robertDouglass, pwolanin | lazysoundsystem: Fixed Typo in apachesolr_search().module.
#763072 by robertDouglass, justinrandell | pwolanin: Fixed warnings when indexing old, crappy html.
#658278 by cpliakas | JThan: Fixed Errors when building the search index in PHP 5.3.
#765486 by robertDouglass: Fixed Several cases where Luke cache not getting cleared and resulting in errors.
#765448 by robertDouglass: Fixed Facet blocks for hierarchical taxonomy broken.
#751420 by pwolanin, skwashd | Damien Tournoud, Scott Reynolds: Fixed apachesolr_site_hash() calls md5() twice.
#750426 by mkalkbrenner, pwolanin | robertDouglass: Fixed fieldType textTight conficts with fieldType text and textSpell.

Apache Solr integration 6.x-2.0-BETA1, 2010-03-24
------------------------------
#649038 by brunodbo, slip | robertDouglass: Fixed Search not working on 404 page.

Apache Solr integration 6.x-2.0-ALPHA3, 2010-03-22
------------------------------
#610656 by pwolanin, claudiu.cristea | Scott Reynolds: Fixed Facets requests for non-enabled modules.
#686390 by pwolanin | rjbrown99: Fixed Wrong number of initial items in taxonomy facet under certain conditions.
#573734 by drewish | robertDouglass: Added Index controls should be radio buttons with one form submission button.
#736540 by drewish | Scott Reynolds: Changed Minimize UPDATE queries in apachesolr_nodeapi_mass_update().
#687738 by David Lesieur | anantagati: Fixed Avoid introducing empty 'filters' query string.
#733116 by pwolanin | drewish: Changed Implement hook_flush_caches().
#719356 by robertDouglass, mathieu | flk: Fixed Indexing cron triggers sigsegv in apachesolr.module line 387.
#744038 by siliconmeadow: Changed Change of Drush extension command naming conventions.
#558160 by robertDouglass, mihha | DenRaf, mcarbone, haxney: Added date facet for cck fields.
#666936 by pwolanin, robertDouglass, claudiu.cristea | justindodge: Fixed apachesolr.js - Drupal.behaviors.apachesolr does not respect context.
#708424 by janusman: Changed Change gmdate() to Drupal format_date() in date facets to support localization.

Apache Solr integration 6.x-2.0-ALPHA2, 2010-01-08
------------------------------
#679522 by pwolanin, Add gettableFiles to solr admin interface config.
http://drupal.org/cvs?commit=309746 by robertDouglass, add entity='comment' to comments on indexing.
#672882 by David Lesieur: Fixed Broken 'Show more' link on taxonomy facets.
#604566 by robertDouglass | jhedstrom: Fixed index_value() never set for CCK fields that aren't of type text, node or user referrence.
#672530 by robertDouglass: Fixed Change array key names from display callback to display_callback() and indexing callback to indexing_callback().
#672518 by robertDouglass: Fixed Add new trie prefixes to helper function .
#551582 by drewish: Fixed Show value instead of key in CCK facets.
#668396 by pwolanin, closer to fix from #655006 for PHP notices.
#664818 by robertDouglass, pounard, pwolanin | Scott Reynolds: Fixed Wrong watchdog() usage.

apachesolr.js
// $Id: apachesolr.js,v 1.5 2010/05/12 20:21:38 pwolanin Exp $
(function ($) {

Drupal.behaviors.apachesolr = {
  attach: function(context, settings) {
    $('.apachesolr-hidden-facet', context).hide();
    $('<a href="#" class="apachesolr-showhide"></a>').text(Drupal.t('Show more')).click(function() {
      if ($(this).parent().find('.apachesolr-hidden-facet:visible').length == 0) {
        $(this).parent().find('.apachesolr-hidden-facet').show();
        $(this).text(Drupal.t('Show fewer'));
      }
      else {
        $(this).parent().find('.apachesolr-hidden-facet').hide();
        $(this).text(Drupal.t('Show more'));
      }
      return false;
    }).appendTo($(settings.apachesolr_show_more_blocks, context));

    if (settings.apachesolr_facetstyle == 'checkboxes') {
      // Find all facet links and give them a checkbox
      $('.apachesolr-facet', context).each(Drupal.apachesolr.addCheckbox);
      // Find all unclick links and turn them into checkboxes
      $('.apachesolr-unclick', context).each(Drupal.apachesolr.makeCheckbox);
    }
  }
}

Drupal.apachesolr = {}

Drupal.apachesolr.addCheckbox = function() {
  // Put href in context scope to be visible in the anonymous function.
  var href = $(this).attr('href');
  $(this).before($('<input type="checkbox" />')
    .attr('class', 'facet-checkbox')
    .click(function(){
      window.location.href = href;
    })
  );
}

Drupal.apachesolr.makeCheckbox = function() {
  // Create a checked checkbox.
  var checkbox = $('<input type="checkbox" />')
    .attr('class', 'facet-checkbox')
    .attr('checked', true);
  // Put href in context scope to be visible in the anonymous function.
  var href = $(this).attr('href');
  checkbox.click(function(){
    window.location.href = href;
  });
  // Add the checkbox, hide the link.
  $(this).before(checkbox).hide();
}

})(jQuery);

Drupal_Apache_Solr_Service.php
<?php
// $Id: Drupal_Apache_Solr_Service.php,v 1.12 2010/08/13 21:16:22 pwolanin Exp $

require_once 'SolrPhpClient/Apache/Solr/Service.php';

class Drupal_Apache_Solr_Service extends Apache_Solr_Service {

  protected $luke;
  protected $luke_cid;
  protected $stats;
  const LUKE_SERVLET = 'admin/luke';
  const STATS_SERVLET = 'admin/stats.jsp';

  /**
   * Call the /admin/ping servlet, to test the connection to the server.
   *
   * @param $timeout
   *   maximum time to wait for ping in seconds, -1 for unlimited (default 2).
   * @return
   *   (float) seconds taken to ping the server, FALSE if timeout occurs.
   */
  public function ping($timeout = 2) {
    $start = microtime(TRUE);

    if ($timeout <= 0.0) {
      $timeout = -1;
    }
    // Attempt a HEAD request to the solr ping url.
    list($data, $headers) = $this->_makeHttpRequest($this->_pingUrl, 'HEAD', array(), NULL, $timeout);
    $response = new Apache_Solr_Response($data, $headers);

    if ($response->getHttpStatus() == 200) {
      // Add 0.1 ms to the ping time so we never return 0.0.
      return microtime(TRUE) - $start + 0.0001;
    }
    else {
      return FALSE;
    }
  }

  /**
   * Sets $this->luke with the meta-data about the index from admin/luke.
   */
  protected function setLuke($num_terms = 0) {
    if (empty($this->luke[$num_terms])) {
      $url = $this->_constructUrl(self::LUKE_SERVLET, array('numTerms' => "$num_terms", 'wt' => self::SOLR_WRITER));
      $this->luke[$num_terms] = $this->_sendRawGet($url);
      cache_set($this->luke_cid, $this->luke, 'cache_apachesolr');
    }
  }

  /**
   * Get just the field meta-data about the index.
   */
  public function getFields($num_terms = 0) {
    return $this->getLuke($num_terms)->fields;
  }

  /**
   * Get meta-data about the index.
   */
  public function getLuke($num_terms = 0) {
    if (!isset($this->luke[$num_terms])) {
      $this->setLuke($num_terms);
    }
    return $this->luke[$num_terms];
  }

  /**
   * Sets $this->stats with the information about the Solr Core form /admin/stats.jsp
   */
  protected function setStats() {
    $data = $this->getLuke();
    // Only try to get stats if we have connected to the index.
    if (empty($this->stats) && isset($data->index->numDocs)) {
      $url = $this->_constructUrl(self::STATS_SERVLET);
      $this->stats_cid = "apachesolr:stats:" . md5($url);
      $cache = cache_get($this->stats_cid, 'cache_apachesolr');
      if (isset($cache->data)) {
        $this->stats = simplexml_load_string($cache->data);
      }
      else {
        $response = $this->_sendRawGet($url);
        $this->stats = simplexml_load_string($response->getRawResponse());
        cache_set($this->stats_cid, $response->getRawResponse(), 'cache_apachesolr');
      }
    }
  }

  /**
   * Get information about the Solr Core.
   *
   * Returns a Simple XMl document
   */
  public function getStats() {
    if (!isset($this->stats)) {
      $this->setStats();
    }
    return $this->stats;
  }

  /**
   * Get summary information about the Solr Core.
   */
  public function getStatsSummary() {
    $stats = $this->getStats();
    $summary = array(
     '@pending_docs' => '',
     '@autocommit_time_seconds' => '',
     '@autocommit_time' => '',
     '@deletes_by_id' => '',
     '@deletes_by_query' => '',
     '@deletes_total' => '',
     '@schema_version' => '',
     '@core_name' => '',
    );

    if (!empty($stats)) {
      $docs_pending_xpath = $stats->xpath('//stat[@name="docsPending"]');
      $summary['@pending_docs'] = (int) trim($docs_pending_xpath[0]);
      $max_time_xpath = $stats->xpath('//stat[@name="autocommit maxTime"]');
      $max_time = (int) trim(current($max_time_xpath));
      // Convert to seconds.
      $summary['@autocommit_time_seconds'] = $max_time / 1000;
      $summary['@autocommit_time'] = format_interval($max_time / 1000);
      $deletes_id_xpath = $stats->xpath('//stat[@name="deletesById"]');
      $summary['@deletes_by_id'] = (int) trim($deletes_id_xpath[0]);
      $deletes_query_xpath = $stats->xpath('//stat[@name="deletesByQuery"]');
      $summary['@deletes_by_query'] = (int) trim($deletes_query_xpath[0]);
      $summary['@deletes_total'] = $summary['@deletes_by_id'] + $summary['@deletes_by_query'];
      $schema = $stats->xpath('/solr/schema[1]');
      $summary['@schema_version'] = trim($schema[0]);;
      $core = $stats->xpath('/solr/core[1]');
      $summary['@core_name'] = trim($core[0]);
    }

    return $summary;
  }

  /**
   * Clear cached Solr data.
   */
  public function clearCache() {
    // Don't clear cached data if the server is unavailable.
    if (@$this->ping()) {
      $this->_clearCache();
    }
    else {
      throw new Exception('No Solr instance available when trying to clear the cache.');
    }
  }

  protected function _clearCache() {
    cache_clear_all("apachesolr:luke:", 'cache_apachesolr', TRUE);
    cache_clear_all("apachesolr:stats:", 'cache_apachesolr', TRUE);
    $this->luke = array();
    $this->stats = NULL;
  }

  /**
   * Clear the cache whenever we commit changes.
   *
   * @see Apache_Solr_Service::commit()
   */
  public function commit($optimize = TRUE, $waitFlush = TRUE, $waitSearcher = TRUE, $timeout = 3600) {
    parent::commit($optimize, $waitFlush, $waitSearcher, $timeout);
    $this->_clearCache();
  }

  /**
   * Construct the Full URLs for the three servlets we reference
   *
   * @see Apache_Solr_Service::_initUrls()
   */
  protected function _initUrls() {
    parent::_initUrls();
    $this->_lukeUrl = $this->_constructUrl(self::LUKE_SERVLET, array('numTerms' => '0', 'wt' => self::SOLR_WRITER));
  }

  /**
   * Make a request to a servlet (a path) that's not a standard path.
   *
   * @param string $servlet
   *   A path to be added to the base Solr path. e.g. 'extract/tika'
   *
   * @param array $params
   *   Any request parameters when constructing the URL.
   *
   * @param string $method
   *   'GET', 'POST', 'PUT', or 'HEAD'.
   *
   * @param array $request_headers
   *   Keyed array of header names and values.  Should include 'Content-Type'
   *   for POST or PUT.
   *
   * @param string $rawPost
   *   Must be an empty string unless method is POST or PUT.
   *
   * @param float $timeout
   *   Read timeout in seconds or FALSE.
   *
   * @return
   *  Apache_Solr_Response object
   */
  public function makeServletRequest($servlet, $params = array(), $method = 'GET', $request_headers = array(), $rawPost = '', $timeout = FALSE) {
    if ($method == 'GET' || $method == 'HEAD') {
      // Make sure we are not sending a request body.
      $rawPost = '';
    }
    // Add default params.
    $params += array(
      'wt' => self::SOLR_WRITER,
    );

    $url = $this->_constructUrl($servlet, $params);
    list($data, $headers) = $this->_makeHttpRequest($url, $method, $request_headers, $rawPost, $timeout);
    $response = new Apache_Solr_Response($data, $headers, $this->_createDocuments, $this->_collapseSingleValueArrays);
    $code = (int) $response->getHttpStatus();
    if ($code != 200) {
      $message = $response->getHttpStatusMessage();
      if ($code >= 400 && $code != 403 && $code != 404) {
        // Add details, like Solr's exception message.
        $message .= $response->getRawResponse();
      }
      throw new Exception('"' . $code . '" Status: ' . $message);
    }
    return $response;
  }

  /**
   * Put Luke meta-data from the cache into $this->luke when we instantiate.
   *
   * @see Apache_Solr_Service::__construct()
   */
  public function __construct($host = 'localhost', $port = 8180, $path = '/solr/') {
    parent::__construct($host, $port, $path);
    $this->luke_cid = "apachesolr:luke:" . md5($this->_lukeUrl);
    $cache = cache_get($this->luke_cid, 'cache_apachesolr');
    if (isset($cache->data)) {
      $this->luke = $cache->data;
    }
  }

  /**
   * Central method for making a get operation against this Solr Server
   *
   * @see Apache_Solr_Service::_sendRawGet()
   */
  protected function _sendRawGet($url, $timeout = FALSE) {
    list($data, $headers) = $this->_makeHttpRequest($url, 'GET', array(), '', $timeout);
    $response = new Apache_Solr_Response($data, $headers, $this->_createDocuments, $this->_collapseSingleValueArrays);
    $code = (int) $response->getHttpStatus();
    if ($code != 200) {
      $message = $response->getHttpStatusMessage();
      if ($code >= 400 && $code != 403 && $code != 404) {
        // Add details, like Solr's exception message.
        $message .= $response->getRawResponse();
      }
      throw new Exception('"' . $code . '" Status: ' . $message);
    }
    return $response;
  }

  /**
   * Central method for making a post operation against this Solr Server
   *
   * @see Apache_Solr_Service::_sendRawPost()
   */
  protected function _sendRawPost($url, $rawPost, $timeout = FALSE, $contentType = 'text/xml; charset=UTF-8') {
    $request_headers = array('Content-Type' => $contentType);
    list($data, $headers) = $this->_makeHttpRequest($url, 'POST', $request_headers, $rawPost, $timeout);
    $response = new Apache_Solr_Response($data, $headers, $this->_createDocuments, $this->_collapseSingleValueArrays);
    $code = (int) $response->getHttpStatus();
    if ($code != 200) {
      $message = $response->getHttpStatusMessage();
      if ($code >= 400 && $code != 403 && $code != 404) {
        // Add details, like Solr's exception message.
        $message .= $response->getRawResponse();
      }
      throw new Exception('"' . $code . '" Status: ' . $message);
    }
    return $response;
  }

  protected function _makeHttpRequest($url, $method = 'GET', $headers = array(), $content = '', $timeout = FALSE) {
    $options = array(
      'headers' => $headers,
      'method' => $method,
      'data' => $content
    );

    if ($timeout) {
      $options['timeout'] = $timeout;
    }

    $result = drupal_http_request($url, $options);

    if (!isset($result->code) || $result->code < 0) {
      $result->code = 0;
      $result->status_message = 'Request failed';
    }
    // Additional information may be in the error property.
    if (isset($result->error)) {
      $result->status_message .= ': ' . check_plain($result->error);
    }

    if (!isset($result->data)) {
      $result->data = '';
    }
    // The headers have to be reformatted for the response class.
    $headers[] = "{$result->protocol} {$result->code} {$result->status_message}";
    if (isset($result->headers)) {
      foreach ($result->headers as $name => $value) {
        $headers[] = "$name: $value";
      }
    }
    return array($result->data, $headers);
  }
}
